src/Libraries/Microsoft.Extensions.AI/ChatCompletion/ChatClientBuilderServiceCollectionExtensions.cs

```diff
@@ -11,37 +11,71 @@ namespace Microsoft.Extensions.DependencyInjection;
+    /// <param name="innerClientFactory">A callback that produces the inner <see cref="IChatClient"/> that represents the underlying backend.</param>
+    /// <returns>A <see cref="ChatClientBuilder"/> that can be used to build a pipeline around the inner client.</returns>
+    /// <remarks>The client is registered as a singleton service.</remarks>
+    public static ChatClientBuilder AddChatClient(
+        this IServiceCollection serviceCollection,
+        Func<IServiceProvider, IChatClient> innerClientFactory)
     {
-        _ = Throw.IfNull(services);
-        _ = Throw.IfNull(clientFactory);
+        _ = Throw.IfNull(serviceCollection);
+        _ = Throw.IfNull(innerClientFactory);
 
-        return services.AddScoped(services =>
-            clientFactory(new ChatClientBuilder(services)));
+        var builder = new ChatClientBuilder(innerClientFactory);
+        _ = serviceCollection.AddSingleton(builder.Build);
+        return builder;
     }
 
     /// <summary>Adds a chat client to the <see cref="IServiceCollection"/>.</summary>
-    /// <param name="services">The <see cref="IServiceCollection"/> to which the client should be added.</param>
+    /// <param name="serviceCollection">The <see cref="IServiceCollection"/> to which the client should be added.</param>
+    /// <param name="serviceKey">The key with which to associate the client.</param>
+    /// <param name="innerClient">The inner <see cref="IChatClient"/> that represents the underlying backend.</param>
+    /// <returns>A <see cref="ChatClientBuilder"/> that can be used to build a pipeline around the inner client.</returns>
+    /// <remarks>The client is registered as a scoped service.</remarks>
+    public static ChatClientBuilder AddKeyedChatClient(
+        this IServiceCollection serviceCollection,
+        object serviceKey,
+        IChatClient innerClient)
+        => AddKeyedChatClient(serviceCollection, serviceKey, _ => innerClient);
+
+    /// <summary>Adds a chat client to the <see cref="IServiceCollection"/>.</summary>
+    /// <param name="serviceCollection">The <see cref="IServiceCollection"/> to which the client should be added.</param>
     /// <param name="serviceKey">The key with which to associate the client.</param>
-    /// <param name="clientFactory">The factory to use to construct the <see cref="IChatClient"/> instance.</param>
-    /// <returns>The <paramref name="services"/> collection.</returns>
+    /// <param name="innerClientFactory">A callback that produces the inner <see cref="IChatClient"/> that represents the underlying backend.</param>
+    /// <returns>A <see cref="ChatClientBuilder"/> that can be used to build a pipeline around the inner client.</returns>
     /// <remarks>The client is registered as a scoped service.</remarks>
-    public static IServiceCollection AddKeyedChatClient(
-        this IServiceCollection services,
+    public static ChatClientBuilder AddKeyedChatClient(
```

Comment:
Did you explicitly choose not to have the generic version as well that you added for `AddChatClient<T>`?

Reply:
At the time yes it was intentional, but I'm glad you challenged this because my reasoning didn't actually hold up on closer inspection. I've now added this, but am [leaning towards removing them both](https://github.com/dotnet/extensions/pull/5642#pullrequestreview-2434998332).